cou/apps/ceph.py

```diff
@@ -0,0 +1,82 @@
+# Copyright 2023 Canonical Limited
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Ceph application class."""
+import logging
+from typing import Optional
+
+from cou.apps.app import AppFactory
+from cou.apps.auxiliary import OpenStackAuxiliaryApplication
+from cou.steps import UpgradeStep
+from cou.utils.openstack import OpenStackRelease
+
+logger = logging.getLogger(__name__)
+
+
+@AppFactory.register_application(["ceph-mon"])
+class CephMonApplication(OpenStackAuxiliaryApplication):
+    """Application for Ceph Monitor charm."""
+
+    def pre_upgrade_plan(self, target: OpenStackRelease) -> list[Optional[UpgradeStep]]:
+        """Pre Upgrade planning.
+
+        :param target: OpenStack release as target to upgrade.
+        :type target: OpenStackRelease
+        :return: Plan that will add pre upgrade as sub steps.
+        :rtype: list[Optional[UpgradeStep]]
+        """
+        return [
+            self._get_upgrade_current_release_packages_plan(),
+            self._get_refresh_charm_plan(target),
+            self._get_change_require_osd_release_plan(target, self.expected_current_channel),
+        ]
+
+    def post_upgrade_plan(self, target: OpenStackRelease) -> list[Optional[UpgradeStep]]:
+        """Post Upgrade planning.
+
+        :param target: OpenStack release as target to upgrade.
+        :type target: OpenStackRelease
+        :return: Plan that will add post upgrade as sub steps.
+        :rtype: list[Optional[UpgradeStep]]
+        """
+        return [
+            self._get_reached_expected_target_plan(target),
+            self._get_change_require_osd_release_plan(target, self.target_channel(target)),
+        ]
+
+    def _get_change_require_osd_release_plan(
+        self, target: OpenStackRelease, channel: str, parallel: bool = False
+    ) -> Optional[UpgradeStep]:
+        """Get plan to set correct value for require-osd-release option on ceph-mon.
+
+        :param target: OpenStack release as target to upgrade.
+        :type target: OpenStackRelease
+        :param channel: The channel to get ceph track from.
+        :type str
+        :param parallel: Parallel running, defaults to False
+        :type parallel: bool, optional
+        :return: Plan to check if application workload has been upgraded
+        :rtype: UpgradeStep
+        """
+        if self.expected_current_channel != self.target_channel(target):
```

Comment:
Imagine we get a cloud that it's on `ussuri` but for some reason they didn't update `require-osd-release` and it's as `natilus`. We are trying to upgrade to Victoria but this logic won't let update to octupus because the expected_current_channel and target_channel are equal.

I think this logic should be on `post_upgrade_plan` to decide or not calling ` _get_change_require_osd_release_plan`.

Do you know if there are consequences when calling `require-osd-release` as for example restarting the services? If we do have, maybe we should query the value first before calling it to change. If there is no consequences, then we should always call it on `pre_upgrade_plan` just to ensure that it's set right.

Reply:
As discussed elsewhere, overwriting with the same value doesn't seem to have consequences. However, after careful consideration, check its value before setting is the safest and best practices. I have added this check in the step's function instead of a plan function because it requires async